File: alds1/4c_search3_trie.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const (
	A = iota
	C
	G
	T
	TERM
)

type node struct {
	child [5]*node
}

type Hash struct {
	root *node
}

func NewHash() *Hash {
	h := &Hash{nil}
	return h
}

func (h *Hash) Insert(s string) {
	var rec func(nd *node, s string) *node
	rec = func(nd *node, s string) *node {
		if nd == nil {
			nd = &node{}
		}
		if s == "" {
			nd.child[TERM] = &node{}
			return nd
		}
		i := strings.Index("ACGT", s[:1])
		nd.child[i] = rec(nd.child[i], s[1:])
		return nd
	}
	h.root = rec(h.root, s)
}

func (h *Hash) Find(s string) bool {
	var rec func(nd *node, s string) bool
	rec = func(nd *node, s string) bool {
		if nd == nil {
			return false
		}
		if s == "" {
			if nd.child[TERM] != nil {
				return true
			} else {
				return false
			}
		} else {
			i := strings.Index("ACGT", s[:1])
			return rec(nd.child[i], s[1:])
		}
	}
	return rec(h.root, s)
}

func main() {
	r := bufio.NewScanner(os.Stdin)
	r.Scan()
	n, _ := strconv.Atoi(r.Text())
	h := NewHash()
	for i := 0; i < n; i++ {
		r.Scan()
		args := strings.Fields(r.Text())
		switch args[0] {
		case "insert":
			h.Insert(args[1])
		case "find":
			if h.Find(args[1]) {
				fmt.Println("yes")
			} else {
				fmt.Println("no")
			}
		default:
			fmt.Fprintf(os.Stderr, "unknown command\n")
			os.Exit(1)
		}
	}
}

File: alds1/4c_search3_trie.c
#include <stdio.h>
 
int parent(int i) {
    if (i) return (i - 1) /4;
    else return 0;
}
 
int child(int i, char c) {
    switch (c) {
    case 'A':
        return i*4 + 1;
        break;
    case 'C':
        return i*4 + 2;
        break;
    case 'G':
        return i*4 + 3;
        break;
    case 'T':
        return i*4 + 4;
        break;
    default:
        return -1;
    }
}
 
unsigned char H[2100000];
 
int index(int i) {
    return i / 8;
}
 
int offset(int i) {
    return i % 8;
}
 
void Set(int i) {
    int j = index(i);
    int off = offset(i);
    H[j] = H[j] | (1 << off);
}
 
int Get(int i) {
    int j = index(i);
    int off = offset(i);
    return H[j] & (1 << off);
}
 
void Insert(const char *s) {
    int nd = 0;
    const char *p;
    for (p = s; *p != '\0'; p++) {
        nd = child(nd, *p);
    }
    Set(nd);
}
 
int Find(const char *s) {
    int nd = 0;
    const char *p;
    for (p = s; *p != '\0'; p++) {
        nd = child(nd, *p);
    }
    return Get(nd);
}
 
int main(int argc, char **argv)
{
    int n, i;
    char s[32];
    scanf("%d", &n);
    for (i = 0; i < n; i++) {
        scanf("%s", s);
        if (s[0] == 'i') {
            scanf("%s", s);
            Insert(s);
        } else {
            scanf("%s", s);
            if (Find(s)) {
                printf("yes\n");
            } else {
                printf("no\n");
            }
        }
    }
 
    return 0;
}